artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java

```diff
@@ -228,11 +228,34 @@ private void configureSizeMetric() {
       size.setMax(maxSize, maxSize, maxMessages, maxMessages);
    }
 
+   private boolean validateNewSettings(final AddressSettings addressSettings) {
+
+      Long newPageLimitBytes = addressSettings.getPageLimitBytes();
+
+      if (newPageLimitBytes != null && newPageLimitBytes.longValue() < 0) {
+         newPageLimitBytes = null;
+      }
+      int newPageSize = storageManager.getAllowedPageSize(addressSettings.getPageSizeBytes());
+      if (newPageLimitBytes != null && newPageSize > 0) {
+         long newEstimatedMaxPages = newPageLimitBytes / newPageSize;
+         if (this.numberOfPages > newEstimatedMaxPages) {
+            ActiveMQServerLogger.LOGGER.pageSettingsFailedApply(storeName, addressSettings, "estimated max pages " + newEstimatedMaxPages + " is less than current number of pages " + this.numberOfPages);
+            return false;
+         }
+      }
+      return true;
+   }
+
    /**
     * @param addressSettings
     */
    @Override
    public void applySetting(final AddressSettings addressSettings) {
+
+      if (!validateNewSettings(addressSettings)) {
+         return;
+      }
```

Comment:
The containing applySetting method is also called from the constructor, but this could make it 'silently' (not counting the logging, which is out of band) abort and yet successfully return without apply _any_ of the settings, and so perhaps just continue on with entirely the wrong settings. That feels off.

Reply:
hmm, that's ugly. I think we can do
1. make sure pageLimitBytes is greater than or equal pageSize (i.e. at least one page file should be allowed)
2. if page store is not running (running flag is false), check the number of pages in the store and compare the new max estimated pages, and return false if the number of the new estimated pages is smaller than the number of existing page files.

Reply:
The limit > page size requirement would remove any issue with e.g 0 page files at the low end. Though essentially the same thing occurs with `n` files at the top end, unless the limit is always a multiple of the page size.

The main issue here is that the limit is tracked by a file count based on current limit/size, and so any existing 'old' files of the 'wrong' size can over/underutilise the effective limit if you adjust the page size with outstanding page files. There doesnt seem to be a way around that without byte-level tracking, which is likely to be awkward.

I think we should perhaps just better document that the effective limit is a file count based on current limit/page size, and so changing page size whilst there are existing outstanding pages can lead to such behaviour and if you want to influence things to e.g unblock earlier it might be necessary to adjust the limit at the same time whilst those old files are present.

Reply:
Another thing I realised is that the page size is adjusted further down this method for the JDBC use case. So validating the value at the start of this method, before that later change occurs, is likely to give the wrong behaviour.

Reply:
I've added test to cover the outstanding page files ( > 1). Internally we use number of page files to calculate the limit against the limit which is specified by the bytes. I guess that is fine as long as the user understands and use them properly. I'll add some document for that.

Regarding jdbc I haven't noticed the limit also apply to database store. I'll take a look and see what impact would it be.

Reply:
I dont know how this interacts with the JDBC bits, I just noticed the fact that it messes around with the effective pageSize later in this applySettings method. However I hadnt noticed it was updating a field, and the validateNewSettings() method doesnt use that does calculates the same value locally..so I was probably wrong to be concerned about that bit.

Reply:
The validation uses storageManager.getAllowedPageSize(addressSettings.getPageSizeBytes()) to get the effective pageSizeBytes so that would take care of the jdbc case. Anyway the jdbc bit only is only used to store the page bytes and is not involved in the calculation which is done all in pagingStoreImpl.

Reply:
Yep, the JDBC bit was based on a mis-read.